<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Computing Cepo cell identity genes — Cepo • Cepo</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Computing Cepo cell identity genes — Cepo"><meta property="og:description" content="ExprsMat accepts various matrix objects,
including DelayedArray and HDF5Array for
out-of-memory computations. See vignette."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Cepo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/cepo.html">Cepo: a method for differential stability analysis of scRNA-seq data</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Computing Cepo cell identity genes</h1>
    
    <div class="hidden name"><code>Cepo.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>ExprsMat accepts various matrix objects,
including DelayedArray and HDF5Array for
out-of-memory computations. See vignette.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">Cepo</span><span class="op">(</span>
  <span class="va">exprsMat</span>,
  <span class="va">cellTypes</span>,
  minCells <span class="op">=</span> <span class="fl">20</span>,
  minCelltype <span class="op">=</span> <span class="fl">3</span>,
  exprsPct <span class="op">=</span> <span class="cn">NULL</span>,
  prefilter_sd <span class="op">=</span> <span class="cn">NULL</span>,
  prefilter_pzero <span class="op">=</span> <span class="cn">NULL</span>,
  logfc <span class="op">=</span> <span class="cn">NULL</span>,
  computePvalue <span class="op">=</span> <span class="cn">NULL</span>,
  computeFastPvalue <span class="op">=</span> <span class="cn">TRUE</span>,
  variability <span class="op">=</span> <span class="st">"CV"</span>,
  method <span class="op">=</span> <span class="st">"weightedMean"</span>,
  weight <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.5</span><span class="op">)</span>,
  workers <span class="op">=</span> <span class="fl">1L</span>,
  block <span class="op">=</span> <span class="cn">NULL</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>exprsMat</dt>
<dd><p>Expression matrix where columns denote cells and rows
denote genes</p></dd>
<dt>cellTypes</dt>
<dd><p>Vector of cell type labels</p></dd>
<dt>minCells</dt>
<dd><p>Integer indicating the minimum number of cells required
within a cell type</p></dd>
<dt>minCelltype</dt>
<dd><p>Integer indicating the minimum number of cell types
required in each batch</p></dd>
<dt>exprsPct</dt>
<dd><p>Percentage of lowly expressed genes to remove.
Default to NULL to not remove any genes.</p></dd>
<dt>prefilter_sd</dt>
<dd><p>Numeric value indicating threshold relating to standard
deviation of genes. Used with prefilter_zeros.</p></dd>
<dt>logfc</dt>
<dd><p>Numeric value indicating the threshold of log fold-change
to use to filter genes.</p></dd>
<dt>computePvalue</dt>
<dd><p>Whether to compute p-values using bootstrap test.
Default to NULL to not make computations.
Set this to an integer to set the number of bootstraps needed
(recommend to be at least 100).</p></dd>
<dt>computeFastPvalue</dt>
<dd><p>Logical vector indicating whether to perform a
faster version of p-value calculation. Set to TRUE by default.</p></dd>
<dt>variability</dt>
<dd><p>A character indicating the stability measure
(CV, IQR, MAD, SD). Default is set to CV.</p></dd>
<dt>method</dt>
<dd><p>Character indicating the method for integration
the two stability measures. By default this is set to
'weightedMean' with equal weights.</p></dd>
<dt>weight</dt>
<dd><p>Vector of two values indicating the weights for each stability
measure. By default this value is c(0.5, 0.5).</p></dd>
<dt>workers</dt>
<dd><p>Number of cores to use. Default to 1, which invokes
<code><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html" class="external-link">BiocParallel::SerialParam</a></code>.
For workers greater than 1, see the <code>workers</code> argument in
<code><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">BiocParallel::MulticoreParam</a></code> and <code><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">BiocParallel::SnowParam</a></code>.</p></dd>
<dt>block</dt>
<dd><p>Vector of batch labels</p></dd>
<dt>...</dt>
<dd><p>Additional arguments passed to <code><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">BiocParallel::MulticoreParam</a></code>
and <code><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">BiocParallel::SnowParam</a></code>.</p></dd>
<dt>prefilter_pzeros</dt>
<dd><p>Numeric value indicating threshold relating to the
percentage of zero expression of genes. Used with prefilter_sd.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>Returns a list of key genes.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: SummarizedExperiment</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: MatrixGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: matrixStats</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘matrixStats’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:Biobase’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     anyMissing, rowMedians</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘MatrixGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:matrixStats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedMeans, colWeightedMedians, colWeightedSds,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowWeightedSds, rowWeightedVars</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:Biobase’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     rowMedians</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomicRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomeInfoDb</span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">'cellbench'</span>, package <span class="op">=</span> <span class="st">'Cepo'</span><span class="op">)</span></span>
<span class="r-in"><span class="va">cellbench</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SingleCellExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 894 895 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(3): scPipe Biomart log.exprs.offset</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(2): counts logcounts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(894): AP000902.1 TNNI3 ... SCMH1 IGF2BP2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(895): CELL_000001 CELL_000003 ... CELL_000955 CELL_000965</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(17): unaligned aligned_unmapped ... sizeFactor celltype</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> reducedDimNames(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> mainExpName: NULL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> altExpNames(0):</span>
<span class="r-in"><span class="va">cepoOutput</span> <span class="op">&lt;-</span> <span class="fu">Cepo</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">cellbench</span><span class="op">)</span>, <span class="va">cellbench</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span class="r-in"><span class="va">cepoOutput</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $stats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 889 rows and 3 columns</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                H1975     H2228    HCC827</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AC092447.7  0.849162 -0.448045 -0.401117</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CT45A3      0.831425 -0.406285 -0.425140</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AL049870.3  0.811872 -0.463128 -0.348743</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TDRD9       0.750559 -0.438966 -0.311592</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TNNI3       0.745810 -0.357402 -0.388408</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ...              ...       ...       ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> STK24      -0.654050  0.370391  0.283659</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CPVL       -0.667877  0.138268  0.529609</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BBOX1-AS1  -0.673324  0.437011  0.236313</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> COL4A2     -0.688128  0.397905  0.290223</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> KCNK1      -0.699022  0.330866  0.368156</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $pvalues</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> attr(,"class")</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Cepo" "list"</span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Hani Jieun Kim, Kevin Wang.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

