<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Cepo: a method for differential stability analysis of scRNA-seq data • Cepo</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Cepo: a method for differential stability analysis of scRNA-seq data">
<meta property="og:description" content="Cepo">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Cepo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cepo.html">Cepo: a method for differential stability analysis of scRNA-seq data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="cepo_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Cepo: a method for differential stability analysis of scRNA-seq data</h1>
                        <h4 class="author">Hani Jieun Kim</h4>
            <address class="author_afil">
      The University of Sydney<br><a class="author_email" href="mailto:#"></a><a href="mailto:hani.kim@sydney.edu.au" class="email">hani.kim@sydney.edu.au</a>
      </address>
                  
            <h4 class="date">12/22/2020</h4>
      
      
      <div class="hidden name"><code>cepo.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>We introduce <em>Cepo</em>, a method to determine genes governing cell identity from scRNA-seq data. We propose a biologically motivated metric—differential stability (DS)—to define cell identity. Our motivation is driven by the hypothesis that stable gene expression is a key component of cell identity. This hypothesis implies that genes marking a cell type should be (i) expressed and (ii) stable in its expression relative to other cell types. We translate these criteria into a computational framework where, using predefined cell-type labels, we compute a cell-type-specific score to prioritise genes that are differential stably expressed against other cell types between all cell-type pair comparisons.</p>
<p><em>Cepo</em> is therefore distinct from most methods for differential analysis (e.g., differential expression) that prioritise differences in the mean abundance between cell types. <em>Cepo</em> is able to capture subtle variations in distribution that does not necessarily involve changes in mean. <em>Cepo</em> is particularly suitable for large atlas data as it is computationally efficient and fast. Moreover, <em>Cepo</em> can perform differential stability analysis for multi-group comparisons in single-cell data.</p>
<p>To access the R code used in the vignettes, type:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/browseVignettes.html">browseVignettes</a></span><span class="op">(</span><span class="st">"Cepo"</span><span class="op">)</span></code></pre></div>
<p>Questions relating to <em>Cepo</em> should be reported as a new issue at <em><a href="https://github.com/HaniJieunKim/Cepo/issues">BugReports</a></em>.</p>
<p>To cite <em>Cepo</em>, type:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/citation.html">citation</a></span><span class="op">(</span><span class="st">"Cepo"</span><span class="op">)</span></code></pre></div>
<div id="package-installation" class="section level2">
<h2 class="hasAnchor">
<a href="#package-installation" class="anchor"></a>Package installation</h2>
<p>The development version of <em>Cepo</em> can be installed with the following command:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># install.packages("remotes")</span>
<span class="co"># remotes::install_github("PYangLba/Cepo")</span></code></pre></div>
</div>
</div>
<div id="differential-stability-analysis-using-cepo" class="section level1">
<h1 class="hasAnchor">
<a href="#differential-stability-analysis-using-cepo" class="anchor"></a>Differential stability analysis using Cepo</h1>
<p>The differential stability analysis in <em>Cepo</em> aims to investigate differential stability patterns between cells of different cell types. To use <em>Cepo</em> one needs data with cell type labels (or cluster labels). If no cell-type labels are provided, cells first need to be clustered and classified in groups via some form of clustering algorithms. <em>Cepo</em> can then be applied to identify differentially stable genes between cell types.</p>
<div id="example-data" class="section level2">
<h2 class="hasAnchor">
<a href="#example-data" class="anchor"></a>Example data</h2>
<p>Load the example dataset, a small subset of the <a href="https://github.com/LuyiTian/sc_mixology">Cellbench</a> dataset consisting of 3 cell types 895 cells and 894 genes.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">Cepo</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"cellbench"</span>, package <span class="op">=</span> <span class="st">"Cepo"</span><span class="op">)</span>
<span class="va">cellbench</span>
<span class="co">#&gt; class: SingleCellExperiment </span>
<span class="co">#&gt; dim: 894 895 </span>
<span class="co">#&gt; metadata(3): scPipe Biomart log.exprs.offset</span>
<span class="co">#&gt; assays(2): counts logcounts</span>
<span class="co">#&gt; rownames(894): AP000902.1 TNNI3 ... SCMH1 IGF2BP2</span>
<span class="co">#&gt; rowData names(0):</span>
<span class="co">#&gt; colnames(895): CELL_000001 CELL_000003 ... CELL_000955 CELL_000965</span>
<span class="co">#&gt; colData names(17): unaligned aligned_unmapped ... sizeFactor celltype</span>
<span class="co">#&gt; reducedDimNames(0):</span>
<span class="co">#&gt; altExpNames(0):</span></code></pre></div>
<p>Columns of the <code>colData</code> indicate the individual id and various metadata for each cell. <code>colData</code> contains <code>celltype</code> labels, which will be required to run <em>Cepo</em>. Differential stability analysis performed on the entire cell type repertoire.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">cellbench</span><span class="op">)</span>
<span class="co">#&gt; DataFrame with 895 rows and 17 columns</span>
<span class="co">#&gt;             unaligned aligned_unmapped mapped_to_exon mapped_to_intron</span>
<span class="co">#&gt;             &lt;integer&gt;        &lt;integer&gt;      &lt;integer&gt;        &lt;integer&gt;</span>
<span class="co">#&gt; CELL_000001    167234             8341         526950            40991</span>
<span class="co">#&gt; CELL_000003    174510             8608         513021            42270</span>
<span class="co">#&gt; CELL_000004    158346             7796         504676            39684</span>
<span class="co">#&gt; CELL_000005    159070             6968         486645            38252</span>
<span class="co">#&gt; CELL_000006    144914             8610         465126            33435</span>
<span class="co">#&gt; ...               ...              ...            ...              ...</span>
<span class="co">#&gt; CELL_000939     35877             2227         115822             8543</span>
<span class="co">#&gt; CELL_000943     37475             2229         105134            13754</span>
<span class="co">#&gt; CELL_000946     35066             2039         105214            14480</span>
<span class="co">#&gt; CELL_000955     33183             3061          84351            19273</span>
<span class="co">#&gt; CELL_000965     35006             1634          81662             7936</span>
<span class="co">#&gt;             ambiguous_mapping mapped_to_ERCC mapped_to_MT number_of_genes</span>
<span class="co">#&gt;                     &lt;integer&gt;      &lt;integer&gt;    &lt;integer&gt;       &lt;numeric&gt;</span>
<span class="co">#&gt; CELL_000001             21392              0        22342           11237</span>
<span class="co">#&gt; CELL_000003             20170              0        20943           11203</span>
<span class="co">#&gt; CELL_000004             18628              0        14021           11237</span>
<span class="co">#&gt; CELL_000005             20029              0        14100           10920</span>
<span class="co">#&gt; CELL_000006             21732              0        11855           11157</span>
<span class="co">#&gt; ...                       ...            ...          ...             ...</span>
<span class="co">#&gt; CELL_000939              5184              0         4061            8240</span>
<span class="co">#&gt; CELL_000943              4567              0         5626            7334</span>
<span class="co">#&gt; CELL_000946              3462              0         4050            7315</span>
<span class="co">#&gt; CELL_000955              3319              0         4322            6809</span>
<span class="co">#&gt; CELL_000965              3003              0         2504            7357</span>
<span class="co">#&gt;             total_count_per_cell non_mt_percent non_ribo_percent outliers</span>
<span class="co">#&gt;                        &lt;numeric&gt;      &lt;numeric&gt;        &lt;numeric&gt; &lt;factor&gt;</span>
<span class="co">#&gt; CELL_000001               266880       0.823985         0.797096    FALSE</span>
<span class="co">#&gt; CELL_000003               251204       0.828956         0.801715    FALSE</span>
<span class="co">#&gt; CELL_000004               250040       0.839618         0.816149    FALSE</span>
<span class="co">#&gt; CELL_000005               244441       0.838746         0.798577    FALSE</span>
<span class="co">#&gt; CELL_000006               235288       0.817904         0.788378    FALSE</span>
<span class="co">#&gt; ...                          ...            ...              ...      ...</span>
<span class="co">#&gt; CELL_000939                84182       0.809377         0.811824    FALSE</span>
<span class="co">#&gt; CELL_000943                58540       0.846379         0.773164    FALSE</span>
<span class="co">#&gt; CELL_000946                51903       0.831185         0.834846    FALSE</span>
<span class="co">#&gt; CELL_000955                42862       0.837128         0.796160    FALSE</span>
<span class="co">#&gt; CELL_000965                61664       0.826106         0.791888    FALSE</span>
<span class="co">#&gt;               cell_line cell_line_demuxlet demuxlet_cls sizeFactor    celltype</span>
<span class="co">#&gt;             &lt;character&gt;        &lt;character&gt;  &lt;character&gt;  &lt;numeric&gt; &lt;character&gt;</span>
<span class="co">#&gt; CELL_000001      HCC827             HCC827          SNG    2.15032      HCC827</span>
<span class="co">#&gt; CELL_000003      HCC827             HCC827          SNG    2.00889      HCC827</span>
<span class="co">#&gt; CELL_000004      HCC827             HCC827          SNG    2.06447      HCC827</span>
<span class="co">#&gt; CELL_000005      HCC827             HCC827          SNG    1.92110      HCC827</span>
<span class="co">#&gt; CELL_000006       H1975              H1975          SNG    1.92716       H1975</span>
<span class="co">#&gt; ...                 ...                ...          ...        ...         ...</span>
<span class="co">#&gt; CELL_000939       H1975              H1975          SNG   0.752065       H1975</span>
<span class="co">#&gt; CELL_000943       H1975              H1975          SNG   0.484279       H1975</span>
<span class="co">#&gt; CELL_000946       H2228              H2228          SNG   0.460024       H2228</span>
<span class="co">#&gt; CELL_000955      HCC827             HCC827          SNG   0.380442      HCC827</span>
<span class="co">#&gt; CELL_000965      HCC827             HCC827          SNG   0.572591      HCC827</span></code></pre></div>
<p>Note that, if cell-type labels are unknown, we would need to cluster cells into groups via some clustering algorithm. In the example dataset, we have 3 cell types, H1975, H2228 and HCC827, all of which are commonly used cell lines of lung adenocarcinomas.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">cellbench</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span>
<span class="co">#&gt; [1] "HCC827" "H1975"  "H2228"</span></code></pre></div>
</div>
<div id="run-cepo-to-generate-list-of-cell-identity-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#run-cepo-to-generate-list-of-cell-identity-genes" class="anchor"></a>Run Cepo to generate list of cell identity genes</h2>
<p>There are two main arguments to <em>Cepo</em>: 1) <code>exprsMat</code> is the input data, which should be normalized data, such as counts per million (CPM) or log2-CPM (e.g., <code>logcounts</code> as created via <code>scater::logNormCounts</code>). 2) <code>cellTypes</code> receives as input a vector of cell-type labels. Note that the cell-type labels should be equal in length and ordered the same as the column names in <code>exprsMat</code>.</p>
<p>Lastly, parameter <code>exprs_pct</code> specifies the threshold for filtering of lowly expressed genes should be performed. By default, this is set of <code>NULL</code>. A value between 0 and 1 should be provided. Differential expression analyses have been shown to benefit from filtering of lowly expressed genes. Similarly, for scRNA-seq datasets with many lowly expressed genes, we recommend that the user performs filtering of the data. Whilst there is no set rule to the threshold, we recommend a value between <code>0.05</code> and <code>0.07</code>, which will filter any genes that are not expressed in 5-7% of cells across each cell type.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ds_res</span> <span class="op">=</span> <span class="fu"><a href="../reference/Cepo.html">Cepo</a></span><span class="op">(</span>exprsMat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">logcounts</a></span><span class="op">(</span><span class="va">cellbench</span><span class="op">)</span>,
              cellType <span class="op">=</span> <span class="va">cellbench</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></code></pre></div>
<p>Cepo returns a statistic per cell type, and the resulting statistics can be extracted as a dataframe. A higher value denotes a gene that is more prioritized as a differentially stable gene in that given cell type. The columns denote cell types and the rows denote genes.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ds_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getStats.html">getStats</a></span><span class="op">(</span><span class="va">ds_res</span><span class="op">)</span> 
<span class="co"># alternatively run</span>
<span class="co"># head(ds_res)</span></code></pre></div>
<p>Conversely, we can extract ordered statistics as a list, where each element in the list is an ordered cell-type specific Cepo score.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ds_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getStats.html">getStats</a></span><span class="op">(</span><span class="va">ds_res</span>, returnDF<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">ds_stats</span><span class="op">)</span>
<span class="co">#&gt; List of 3</span>
<span class="co">#&gt;  $ H1975 : Named num [1:889] 0.849 0.831 0.812 0.751 0.746 ...</span>
<span class="co">#&gt;   ..- attr(*, "names")= chr [1:889] "AC092447.7" "CT45A3" "AL049870.3" "TDRD9" ...</span>
<span class="co">#&gt;  $ H2228 : Named num [1:889] 0.859 0.828 0.812 0.795 0.78 ...</span>
<span class="co">#&gt;   ..- attr(*, "names")= chr [1:889] "HLA-DRB6" "AR" "NDN" "DNAJC15" ...</span>
<span class="co">#&gt;  $ HCC827: Named num [1:889] 0.866 0.849 0.784 0.783 0.78 ...</span>
<span class="co">#&gt;   ..- attr(*, "names")= chr [1:889] "CASC9" "AC011632.1" "SPESP1" "AC078962.4" ...</span>
<span class="co">#&gt;  - attr(*, "class")= chr [1:2] "Cepo" "list"</span>
<span class="co">#&gt;  - attr(*, "differential_method")= chr "Cepo"</span></code></pre></div>
<div id="run-differential-expression-analyses" class="section level3">
<h3 class="hasAnchor">
<a href="#run-differential-expression-analyses" class="anchor"></a>Run differential expression analyses</h3>
<p><em>Cepo</em> package contains wrapper functions for several commonly used differential expression analyses methods that the user can use to compare the results with Cepo results. The wrapper functions can be run using the following functions: <code>doLimma</code>, <code>doVoom</code>, <code>doTtest</code> and <code>doWilcoxon</code>. We illustrate the use of <code>doLimma</code> as an example.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">limma_res</span> <span class="op">=</span> <span class="fu"><a href="../reference/doLimma.html">doLimma</a></span><span class="op">(</span>exprsMat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">logcounts</a></span><span class="op">(</span><span class="va">cellbench</span><span class="op">)</span>,
                    cellType <span class="op">=</span> <span class="va">cellbench</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></code></pre></div>
<p>Glimpse of the results.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/glimpse.html">glimpse</a></span><span class="op">(</span><span class="va">limma_res</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span><span class="op">)</span>
<span class="co">#&gt; Rows: 5</span>
<span class="co">#&gt; Columns: 11</span>
<span class="co">#&gt; $ ID          &lt;chr&gt; "AP000902.1", "DHRS2", "ECH1", "TNNI3", "AL049870.3"</span>
<span class="co">#&gt; $ logFC       &lt;dbl&gt; 4.067047, 5.188615, -3.359354, 2.902635, 2.584281</span>
<span class="co">#&gt; $ AveExpr     &lt;dbl&gt; 1.5797165, 2.7830054, 2.5468200, 1.1328379, 0.9427417</span>
<span class="co">#&gt; $ t           &lt;dbl&gt; 121.67189, 70.16253, -65.87735, 62.22522, 57.16111</span>
<span class="co">#&gt; $ P.Value     &lt;dbl&gt; 0.000000e+00, 0.000000e+00, 0.000000e+00, 0.000000e+00, 9…</span>
<span class="co">#&gt; $ adj.P.Val   &lt;dbl&gt; 0.000000e+00, 0.000000e+00, 0.000000e+00, 0.000000e+00, 1…</span>
<span class="co">#&gt; $ B           &lt;dbl&gt; 1289.2070, 838.0223, 790.0330, 747.4892, 685.8754</span>
<span class="co">#&gt; $ meanExprs.1 &lt;dbl&gt; 0.17101848, 0.98583144, 3.71039517, 0.12745610, 0.04762771</span>
<span class="co">#&gt; $ meanExprs.2 &lt;dbl&gt; 4.238066, 6.174447, 0.351041, 3.030091, 2.631909</span>
<span class="co">#&gt; $ meanPct.1   &lt;dbl&gt; 0.15384615, 0.67692308, 0.99829060, 0.11965812, 0.04786325</span>
<span class="co">#&gt; $ meanPct.2   &lt;dbl&gt; 1.0000000, 1.0000000, 0.3096774, 0.9838710, 0.9580645</span></code></pre></div>
<p>As in <code>Cepo</code>, the resulting statistics can be extracted as a dataframe for the differential expression methods as below.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">limma_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getStats.html">getStats</a></span><span class="op">(</span><span class="va">limma_res</span><span class="op">)</span>

<span class="va">limma_stats</span>
<span class="co">#&gt; DataFrame with 889 rows and 3 columns</span>
<span class="co">#&gt;              H1975     H2228      HCC827</span>
<span class="co">#&gt;          &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;</span>
<span class="co">#&gt; A1BG-AS1   18.5735 -18.41675  -0.0891514</span>
<span class="co">#&gt; AADAT      12.6363  -5.21427  -6.9061436</span>
<span class="co">#&gt; ABCC1     -15.3749  33.33666  -9.4036144</span>
<span class="co">#&gt; ABCC2     -17.9499  51.87008 -11.8720560</span>
<span class="co">#&gt; ABCC3     -14.7589  42.79053 -12.7984250</span>
<span class="co">#&gt; ...            ...       ...         ...</span>
<span class="co">#&gt; ZNF91      19.6583  -8.18030   -9.283581</span>
<span class="co">#&gt; ZNF93      17.8269  -8.43805   -7.709299</span>
<span class="co">#&gt; ZNHIT6    -12.3851  -8.36528   27.162069</span>
<span class="co">#&gt; ZSCAN18    19.2075 -18.39443   -0.528456</span>
<span class="co">#&gt; ZSCAN5A    16.8516 -16.50573   -0.238599</span></code></pre></div>
<p>Conversely, we can extract ordered statistics as list for the differential expression methods as below.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">limma_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getStats.html">getStats</a></span><span class="op">(</span><span class="va">limma_res</span>, returnDF <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">limma_stats</span><span class="op">)</span>
<span class="co">#&gt; List of 3</span>
<span class="co">#&gt;  $ H1975 : Named num [1:889] 121.7 70.2 62.2 57.2 54 ...</span>
<span class="co">#&gt;   ..- attr(*, "names")= chr [1:889] "AP000902.1" "DHRS2" "TNNI3" "AL049870.3" ...</span>
<span class="co">#&gt;  $ H2228 : Named num [1:889] 117.5 105.3 105.3 103.5 97.3 ...</span>
<span class="co">#&gt;   ..- attr(*, "names")= chr [1:889] "HLA-DRB1" "HLA-DRB5" "UCHL1" "HLA-DQB1" ...</span>
<span class="co">#&gt;  $ HCC827: Named num [1:889] 87.1 82.9 65.8 63.1 59.6 ...</span>
<span class="co">#&gt;   ..- attr(*, "names")= chr [1:889] "PPP1R14A" "MICOS10P3" "PIP4K2C" "TBK1" ...</span></code></pre></div>
<p>Whilst the default parameter of <code>getStats</code> is to extract the statistics, p-values can be obtained by specifying <code>p-values</code> in <code>parameter_name</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">######################## </span></code></pre></div>
<p>Wrapper functions are available for <code>Limma</code>, <code>Voom</code>, <code>t-test</code> and <code>Wilcoxon</code> test. The code for DE analyses have been adapted from <a href="https://github.com/markrobinsonuzh/conquer">conquer</a>.</p>
</div>
</div>
<div id="visualizing-results" class="section level2">
<h2 class="hasAnchor">
<a href="#visualizing-results" class="anchor"></a>Visualizing results</h2>
<p>We can visualize the overlap of differential stability genes between cell types.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/hms-dbmi/UpSetR">UpSetR</a></span><span class="op">)</span>
<span class="va">res_name</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">ds_res</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">500</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/upset.html">upset</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/fromList.html">fromList</a></span><span class="op">(</span><span class="va">res_name</span><span class="op">)</span>, nsets <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="cepo_files/figure-html/upset-plot-1.png" width="700"></p>
<p>Density plot of two genes from each cell type.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_densities.html">plot_densities</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">cellbench</span>,
               dx_list <span class="op">=</span> <span class="va">ds_res</span>, 
               n_genes <span class="op">=</span> <span class="fl">2</span>,
               assay <span class="op">=</span> <span class="st">"logcounts"</span>, 
               celltype_col <span class="op">=</span> <span class="st">"celltype"</span>,
               celltype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">ds_res</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="cepo_files/figure-html/plot-densities-1.png" width="700"></p>
<p>We can also specify the genes to be plotted.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_densities.html">plot_densities</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">cellbench</span>,
               dx_list<span class="op">=</span><span class="va">ds_res</span>,
               genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"PLTP"</span>, <span class="st">"CPT1C"</span>, <span class="st">"MEG3"</span>, <span class="st">"SYCE1"</span>, <span class="st">"MICOS10P3"</span>, <span class="st">"HOXB7"</span><span class="op">)</span>,
               assay <span class="op">=</span> <span class="st">"logcounts"</span>, 
               celltype_col <span class="op">=</span> <span class="st">"celltype"</span>,
               celltype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">ds_res</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="cepo_files/figure-html/plot-densities-genes-1.png" width="700"></p>
</div>
<div id="running-out-of-memory-computation-with-cepo" class="section level2">
<h2 class="hasAnchor">
<a href="#running-out-of-memory-computation-with-cepo" class="anchor"></a>Running out-of-memory computation with Cepo</h2>
<p>To facilitate analysis of high-throughput atlas data consisting of millions of single cells, Cepo enables out-of-memory and parallel computation.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># BPPARAM = 2</span>
<span class="co"># ds_res_2_cores = Cepo(exprsMat = logcounts(cellbench),</span>
<span class="co">#                      cellTypes = cellbench$celltype,</span>
<span class="co">#                        BPPARAM = BPPARAM)</span></code></pre></div>
</div>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.0.3 (2020-10-10)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Catalina 10.15.7</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] UpSetR_1.4.0                SingleCellExperiment_1.12.0</span>
<span class="co">#&gt;  [3] SummarizedExperiment_1.20.0 Biobase_2.50.0             </span>
<span class="co">#&gt;  [5] GenomicRanges_1.42.0        GenomeInfoDb_1.26.2        </span>
<span class="co">#&gt;  [7] IRanges_2.24.1              S4Vectors_0.28.1           </span>
<span class="co">#&gt;  [9] BiocGenerics_0.36.0         MatrixGenerics_1.2.0       </span>
<span class="co">#&gt; [11] matrixStats_0.57.0          Cepo_0.0.1                 </span>
<span class="co">#&gt; [13] BiocStyle_2.18.1           </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] bitops_1.0-6              fs_1.5.0                 </span>
<span class="co">#&gt;  [3] rprojroot_2.0.2           tools_4.0.3              </span>
<span class="co">#&gt;  [5] backports_1.2.1           utf8_1.1.4               </span>
<span class="co">#&gt;  [7] R6_2.5.0                  colorspace_2.0-0         </span>
<span class="co">#&gt;  [9] gridExtra_2.3             tidyselect_1.1.0         </span>
<span class="co">#&gt; [11] curl_4.3                  compiler_4.0.3           </span>
<span class="co">#&gt; [13] textshaping_0.2.1         cli_2.2.0                </span>
<span class="co">#&gt; [15] desc_1.2.0                DelayedArray_0.16.0      </span>
<span class="co">#&gt; [17] labeling_0.4.2            bookdown_0.21            </span>
<span class="co">#&gt; [19] scales_1.1.1              pkgdown_1.6.1            </span>
<span class="co">#&gt; [21] systemfonts_0.3.2         stringr_1.4.0            </span>
<span class="co">#&gt; [23] digest_0.6.27             foreign_0.8-80           </span>
<span class="co">#&gt; [25] rmarkdown_2.6             rio_0.5.16               </span>
<span class="co">#&gt; [27] XVector_0.30.0            pkgconfig_2.0.3          </span>
<span class="co">#&gt; [29] htmltools_0.5.0           sparseMatrixStats_1.2.0  </span>
<span class="co">#&gt; [31] limma_3.46.0              rlang_0.4.9              </span>
<span class="co">#&gt; [33] readxl_1.3.1              DelayedMatrixStats_1.12.1</span>
<span class="co">#&gt; [35] farver_2.0.3              generics_0.1.0           </span>
<span class="co">#&gt; [37] dplyr_1.0.2               zip_2.1.1                </span>
<span class="co">#&gt; [39] car_3.0-10                RCurl_1.98-1.2           </span>
<span class="co">#&gt; [41] magrittr_2.0.1            GenomeInfoDbData_1.2.4   </span>
<span class="co">#&gt; [43] Matrix_1.2-18             Rcpp_1.0.5               </span>
<span class="co">#&gt; [45] munsell_0.5.0             fansi_0.4.1              </span>
<span class="co">#&gt; [47] abind_1.4-5               lifecycle_0.2.0          </span>
<span class="co">#&gt; [49] stringi_1.5.3             yaml_2.2.1               </span>
<span class="co">#&gt; [51] carData_3.0-4             zlibbioc_1.36.0          </span>
<span class="co">#&gt; [53] plyr_1.8.6                grid_4.0.3               </span>
<span class="co">#&gt; [55] forcats_0.5.0             crayon_1.3.4             </span>
<span class="co">#&gt; [57] lattice_0.20-41           cowplot_1.1.0            </span>
<span class="co">#&gt; [59] haven_2.3.1               splines_4.0.3            </span>
<span class="co">#&gt; [61] hms_0.5.3                 knitr_1.30               </span>
<span class="co">#&gt; [63] pillar_1.4.7              ggpubr_0.4.0             </span>
<span class="co">#&gt; [65] ggsignif_0.6.0            glue_1.4.2               </span>
<span class="co">#&gt; [67] evaluate_0.14             data.table_1.13.4        </span>
<span class="co">#&gt; [69] BiocManager_1.30.10       vctrs_0.3.6              </span>
<span class="co">#&gt; [71] cellranger_1.1.0          gtable_0.3.0             </span>
<span class="co">#&gt; [73] purrr_0.3.4               tidyr_1.1.2              </span>
<span class="co">#&gt; [75] assertthat_0.2.1          ggplot2_3.3.2            </span>
<span class="co">#&gt; [77] xfun_0.19                 openxlsx_4.2.3           </span>
<span class="co">#&gt; [79] broom_0.7.3               rstatix_0.6.0            </span>
<span class="co">#&gt; [81] ragg_0.4.0                tibble_3.0.4             </span>
<span class="co">#&gt; [83] memoise_1.1.0             statmod_1.4.35           </span>
<span class="co">#&gt; [85] ellipsis_0.3.1</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Hani Jieun Kim.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
