<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Cepo: a method for differential stability analysis of scRNA-seq data • Cepo</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Cepo: a method for differential stability analysis of scRNA-seq data">
<meta property="og:description" content="Cepo">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Cepo</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cepo.html">Cepo: a method for differential stability analysis of scRNA-seq data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="cepo_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Cepo: a method for differential stability analysis of scRNA-seq data</h1>
                        <h4 class="author">Hani Jieun Kim</h4>
            <address class="author_afil">
      The University of Sydney<br><a class="author_email" href="mailto:#"></a><a href="mailto:hani.kim@sydney.edu.au" class="email">hani.kim@sydney.edu.au</a>
      </address>
                  
            <h4 class="date">01/12/2021</h4>
      
      
      <div class="hidden name"><code>cepo.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>We introduce <em>Cepo</em>, a method to determine genes governing cell identity from scRNA-seq data. We propose a biologically motivated metric—differential stability (DS)—to define cell identity. Our motivation is driven by the hypothesis that stable gene expression is a key component of cell identity. This hypothesis implies that genes marking a cell type should be (i) expressed and (ii) stable in its expression relative to other cell types. We translate these criteria into a computational framework where, using predefined cell-type labels, we compute a cell-type-specific score to prioritise genes that are differential stably expressed against other cell types between all cell-type pair comparisons.</p>
<p><em>Cepo</em> is therefore distinct from most methods for differential analysis (e.g., differential expression) that prioritise differences in the mean abundance between cell types. <em>Cepo</em> is able to capture subtle variations in distribution that does not necessarily involve changes in mean. <em>Cepo</em> is particularly suitable for large atlas data as it is computationally efficient and fast. Moreover, <em>Cepo</em> can perform differential stability analysis for multi-group comparisons in single-cell data.</p>
<p>To access the R code used in the vignettes, type:</p>
<pre><code><a href="https://rdrr.io/r/utils/browseVignettes.html">browseVignettes("Cepo")</a></code></pre>
<p>Questions relating to <em>Cepo</em> should be reported as a new issue at <em><a href="https://github.com/PYangLab/Cepo/issues">BugReports</a></em>.</p>
<p>To cite <em>Cepo</em>, type:</p>
<pre><code><a href="https://rdrr.io/r/utils/citation.html">citation("Cepo")</a></code></pre>
<div id="package-installation" class="section level2">
<h2 class="hasAnchor">
<a href="#package-installation" class="anchor"></a>Package installation</h2>
<p>The development version of <em>Cepo</em> can be installed with the following command:</p>
<pre><code><a href="https://rdrr.io/r/utils/install.packages.html">install.packages("remotes")
remotes::install_github("PYangLab/Cepo")</a></code></pre>
</div>
</div>
<div id="differential-stability-analysis-using-cepo" class="section level1">
<h1 class="hasAnchor">
<a href="#differential-stability-analysis-using-cepo" class="anchor"></a>Differential stability analysis using Cepo</h1>
<p>The differential stability analysis in <em>Cepo</em> aims to investigate differential stability patterns between cells of different cell types. To use <em>Cepo</em> one needs data with cell type labels (or cluster labels). If no cell-type labels are provided, cells first need to be clustered and classified in groups via some form of clustering algorithms. <em>Cepo</em> can then be applied to identify differentially stable genes between cell types.</p>
<div id="example-data" class="section level2">
<h2 class="hasAnchor">
<a href="#example-data" class="anchor"></a>Example data</h2>
<p>Load the example dataset, a small subset of the <a href="https://github.com/LuyiTian/sc_mixology">Cellbench</a> dataset consisting of 3 cell types 895 cells and 894 genes.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">Cepo</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"cellbench"</span>, package <span class="op">=</span> <span class="st">"Cepo"</span><span class="op">)</span>
<span class="va">cellbench</span></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 894 895 
## metadata(3): scPipe Biomart log.exprs.offset
## assays(2): counts logcounts
## rownames(894): AP000902.1 TNNI3 ... SCMH1 IGF2BP2
## rowData names(0):
## colnames(895): CELL_000001 CELL_000003 ... CELL_000955 CELL_000965
## colData names(17): unaligned aligned_unmapped ... sizeFactor celltype
## reducedDimNames(0):
## altExpNames(0):</code></pre>
<p>Columns of the <code>colData</code> indicate the individual id and various metadata for each cell. <code>colData</code> contains <code>celltype</code> labels, which will be required to run <em>Cepo</em>. Differential stability analysis performed on the entire cell type repertoire.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">colData</span><span class="op">(</span><span class="va">cellbench</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span></code></pre></div>
<pre><code>## DataFrame with 5 rows and 17 columns
##             unaligned aligned_unmapped mapped_to_exon mapped_to_intron
##             &lt;integer&gt;        &lt;integer&gt;      &lt;integer&gt;        &lt;integer&gt;
## CELL_000001    167234             8341         526950            40991
## CELL_000003    174510             8608         513021            42270
## CELL_000004    158346             7796         504676            39684
## CELL_000005    159070             6968         486645            38252
## CELL_000006    144914             8610         465126            33435
##             ambiguous_mapping mapped_to_ERCC mapped_to_MT number_of_genes
##                     &lt;integer&gt;      &lt;integer&gt;    &lt;integer&gt;       &lt;numeric&gt;
## CELL_000001             21392              0        22342           11237
## CELL_000003             20170              0        20943           11203
## CELL_000004             18628              0        14021           11237
## CELL_000005             20029              0        14100           10920
## CELL_000006             21732              0        11855           11157
##             total_count_per_cell non_mt_percent non_ribo_percent outliers
##                        &lt;numeric&gt;      &lt;numeric&gt;        &lt;numeric&gt; &lt;factor&gt;
## CELL_000001               266880       0.823985         0.797096    FALSE
## CELL_000003               251204       0.828956         0.801715    FALSE
## CELL_000004               250040       0.839618         0.816149    FALSE
## CELL_000005               244441       0.838746         0.798577    FALSE
## CELL_000006               235288       0.817904         0.788378    FALSE
##               cell_line cell_line_demuxlet demuxlet_cls sizeFactor    celltype
##             &lt;character&gt;        &lt;character&gt;  &lt;character&gt;  &lt;numeric&gt; &lt;character&gt;
## CELL_000001      HCC827             HCC827          SNG    2.15032      HCC827
## CELL_000003      HCC827             HCC827          SNG    2.00889      HCC827
## CELL_000004      HCC827             HCC827          SNG    2.06447      HCC827
## CELL_000005      HCC827             HCC827          SNG    1.92110      HCC827
## CELL_000006       H1975              H1975          SNG    1.92716       H1975</code></pre>
<p>Note that, if cell-type labels are unknown, we would need to cluster cells into groups via some clustering algorithm. In the example dataset, we have 3 cell types, H1975, H2228 and HCC827, all of which are commonly used cell lines of lung adenocarcinomas.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">cellbench</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "HCC827" "H1975"  "H2228"</code></pre>
</div>
<div id="run-cepo-to-generate-list-of-cell-identity-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#run-cepo-to-generate-list-of-cell-identity-genes" class="anchor"></a>Run Cepo to generate list of cell identity genes</h2>
<p>There are two main arguments to <em>Cepo</em>: 1) <code>exprsMat</code> is the input data, which should be normalized data, such as counts per million (CPM) or log2-CPM (e.g., <code>logcounts</code> as created via <code>scater::logNormCounts</code>). 2) <code>cellTypes</code> receives as input a vector of cell-type labels. Note that the cell-type labels should be equal in length and ordered the same as the column names in <code>exprsMat</code>.</p>
<p>Lastly, parameter <code>exprsPct</code> specifies the threshold for filtering of lowly expressed genes should be performed. By default, this is set of <code>NULL</code>. A value between 0 and 1 should be provided. Differential expression analyses have been shown to benefit from filtering of lowly expressed genes. Similarly, for scRNA-seq datasets with many lowly expressed genes, we recommend that the user performs filtering of the data. Whilst there is no set rule to the threshold, we recommend a value between <code>0.05</code> and <code>0.07</code>, which will filter any genes that are not expressed in 5-7% of cells across each cell type.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ds_res</span> <span class="op">=</span> <span class="fu"><a href="../reference/Cepo.html">Cepo</a></span><span class="op">(</span>exprsMat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">logcounts</a></span><span class="op">(</span><span class="va">cellbench</span><span class="op">)</span>,
              cellType <span class="op">=</span> <span class="va">cellbench</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></code></pre></div>
<p>The <code>Cepo</code> function returns a list of two elements by default. The first element is a <code>DataFrame</code> of DS statistics. In this <code>DataFrame</code>, each column corresponds to the DS statistics for that celltype across all genes. A higher DS statistic value denotes a gene that is more prioritized as a differentially stable gene in that given cell type. In the output DataFrame, the columns correspond to each cell type and each row correspond to a gene.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ds_res</span></code></pre></div>
<pre><code>## $stats
## DataFrame with 889 rows and 3 columns
##                H1975     H2228    HCC827
##            &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## AC092447.7  0.849162 -0.448045 -0.401117
## CT45A3      0.831425 -0.406285 -0.425140
## AL049870.3  0.811872 -0.463128 -0.348743
## TDRD9       0.750559 -0.438966 -0.311592
## TNNI3       0.745810 -0.357402 -0.388408
## ...              ...       ...       ...
## STK24      -0.654050  0.370391  0.283659
## CPVL       -0.667877  0.138268  0.529609
## BBOX1-AS1  -0.673324  0.437011  0.236313
## COL4A2     -0.688128  0.397905  0.290223
## KCNK1      -0.699022  0.330866  0.368156
## 
## $pvalues
## NULL
## 
## attr(,"class")
## [1] "Cepo" "list"</code></pre>
<p>The second element of the <code>Cepo</code> function is a <code>DataFrame</code> of p-values associated with the DS statistics. In this <code>DataFrame</code>, each column corresponds to the p-values associated with the DS statistics obtained through bootstrap on the cells. By default, the <code>Cepo</code> function do not compute these p-values and requires the users to set the <code>computePvalue</code> argument to a number of bootstrap runs required (we recommend this to be at least 100).</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ds_res_pvalues</span> <span class="op">=</span> <span class="fu"><a href="../reference/Cepo.html">Cepo</a></span><span class="op">(</span>exprsMat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">logcounts</a></span><span class="op">(</span><span class="va">cellbench</span><span class="op">)</span>,
              cellType <span class="op">=</span> <span class="va">cellbench</span><span class="op">$</span><span class="va">celltype</span>, computePvalue <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>

<span class="va">ds_res_pvalues</span></code></pre></div>
<pre><code>## $stats
## DataFrame with 889 rows and 3 columns
##                H1975     H2228    HCC827
##            &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## AC092447.7  0.849162 -0.448045 -0.401117
## CT45A3      0.831425 -0.406285 -0.425140
## AL049870.3  0.811872 -0.463128 -0.348743
## TDRD9       0.750559 -0.438966 -0.311592
## TNNI3       0.745810 -0.357402 -0.388408
## ...              ...       ...       ...
## STK24      -0.654050  0.370391  0.283659
## CPVL       -0.667877  0.138268  0.529609
## BBOX1-AS1  -0.673324  0.437011  0.236313
## COL4A2     -0.688128  0.397905  0.290223
## KCNK1      -0.699022  0.330866  0.368156
## 
## $pvalues
## DataFrame with 889 rows and 3 columns
##                H1975     H2228    HCC827
##            &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## AC092447.7         1         0         0
## CT45A3             1         0         0
## AL049870.3         1         0         0
## TDRD9              1         0         0
## TNNI3              1         0         0
## ...              ...       ...       ...
## STK24              0         1         1
## CPVL               0         1         1
## BBOX1-AS1          0         1         1
## COL4A2             0         1         1
## KCNK1              0         1         1
## 
## attr(,"class")
## [1] "Cepo" "list"</code></pre>
<div id="run-differential-expression-analyses" class="section level3">
<h3 class="hasAnchor">
<a href="#run-differential-expression-analyses" class="anchor"></a>Run differential expression analyses</h3>
<p><em>Cepo</em> package contains wrapper functions for several commonly used differential expression analyses methods that the user can use to compare the results with Cepo results. The wrapper functions can be run using the following functions: <code>doLimma</code>, <code>doVoom</code>, <code>doTtest</code> and <code>doWilcoxon</code>. We illustrate the use of <code>doLimma</code> as an example.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">limma_res</span> <span class="op">=</span> <span class="fu"><a href="../reference/doLimma.html">doLimma</a></span><span class="op">(</span>exprsMat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">logcounts</a></span><span class="op">(</span><span class="va">cellbench</span><span class="op">)</span>,
                    cellType <span class="op">=</span> <span class="va">cellbench</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></code></pre></div>
<p>Similar to the output of the <code>Cepo</code> function, the returned output is also a list of two <code>DataFrame</code>s, the first contain the statistic and the second contain the p-values.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">limma_res</span></code></pre></div>
<pre><code>## $stats
## DataFrame with 889 rows and 3 columns
##                H1975     H2228    HCC827
##            &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## AP000902.1  121.6719  -17.8078  -16.3249
## DHRS2        70.1625  -16.9831  -14.8467
## TNNI3        62.2252  -15.8179  -15.1722
## AL049870.3   57.1611  -16.3624  -13.9822
## GSTO2        53.9829  -12.0844  -18.2679
## ...              ...       ...       ...
## PDZK1IP1    -31.9977  39.83966  -2.14271
## COL4A2      -32.0376  24.42532   3.09878
## KCNK1       -37.1582  11.32624  14.77371
## VAMP5       -38.9528   2.48549  32.52836
## ECH1        -65.8773  26.11432   8.31254
## 
## $pvalues
## DataFrame with 889 rows and 3 columns
##                   H1975        H2228       HCC827
##               &lt;numeric&gt;    &lt;numeric&gt;    &lt;numeric&gt;
## AP000902.1  0.00000e+00  5.01783e-61  1.06941e-52
## DHRS2       0.00000e+00  2.40530e-56  8.58973e-45
## ECH1        0.00000e+00 9.19418e-113  3.38733e-16
## TNNI3       0.00000e+00  5.77268e-50  1.69440e-46
## AL049870.3 9.12842e-304  6.14992e-53  2.27145e-40
## ...                 ...          ...          ...
## CRACR2B        0.798799  2.31235e-63  6.53135e-71
## MTAP           0.878480 6.96028e-152 1.59473e-171
## FBXO17         0.944697  2.14798e-62  5.59323e-68
## CCDC74B        0.962182  1.11580e-63  3.01058e-68
## AC093840.1     0.963854  1.41565e-25  1.37280e-27
## 
## attr(,"class")
## [1] "Cepo" "list"</code></pre>
<p>Wrapper functions are available for <code>Limma</code>, <code>Voom</code>, <code>t-test</code> and <code>Wilcoxon</code> test. The code for DE analyses have been adapted from <a href="https://github.com/markrobinsonuzh/conquer">conquer</a>.</p>
</div>
</div>
<div id="visualizing-results" class="section level2">
<h2 class="hasAnchor">
<a href="#visualizing-results" class="anchor"></a>Visualizing results</h2>
<p>We can visualize the overlap of differential stability genes between cell types.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/hms-dbmi/UpSetR">UpSetR</a></span><span class="op">)</span>
<span class="va">res_name</span> <span class="op">=</span> <span class="fu"><a href="../reference/topGenes.html">topGenes</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">ds_res</span>, n <span class="op">=</span> <span class="fl">500</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/upset.html">upset</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/fromList.html">fromList</a></span><span class="op">(</span><span class="va">res_name</span><span class="op">)</span>, nsets <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="cepo_files/figure-html/upset-plot-1.png" width="700"></p>
<p>Density plot of two genes from each cell type.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotDensities.html">plotDensities</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">cellbench</span>,
              cepoOutput <span class="op">=</span> <span class="va">ds_res</span>,
              nGenes <span class="op">=</span> <span class="fl">2</span>,
              assay <span class="op">=</span> <span class="st">"logcounts"</span>,
              celltypeColumn <span class="op">=</span> <span class="st">"celltype"</span><span class="op">)</span></code></pre></div>
<pre><code>## AC092447.7, CT45A3, HLA-DRB6, AR, CASC9, AC011632.1 will be plotted</code></pre>
<p><img src="cepo_files/figure-html/plot-densities-1.png" width="700"></p>
<p>We can also specify the genes to be plotted.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotDensities.html">plotDensities</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">cellbench</span>,
              cepoOutput <span class="op">=</span> <span class="va">ds_res</span>,
              genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"PLTP"</span>, <span class="st">"CPT1C"</span>, <span class="st">"MEG3"</span>, <span class="st">"SYCE1"</span>, <span class="st">"MICOS10P3"</span>, <span class="st">"HOXB7"</span><span class="op">)</span>,
              assay <span class="op">=</span> <span class="st">"logcounts"</span>,
              celltypeColumn <span class="op">=</span> <span class="st">"celltype"</span><span class="op">)</span></code></pre></div>
<p><img src="cepo_files/figure-html/plot-densities-genes-1.png" width="700"></p>
</div>
<div id="running-out-of-memory-computation-with-cepo" class="section level2">
<h2 class="hasAnchor">
<a href="#running-out-of-memory-computation-with-cepo" class="anchor"></a>Running out-of-memory computation with Cepo</h2>
<p>To facilitate analysis of high-throughput atlas data consisting of millions of cells, <code>Cepo</code> also enables out-of-memory and parallel computation.</p>
<p>The <code>Cepo</code> function naturally handles matrices under the <code>DelayedArray</code> wrapper. Briefly, <code>DelayedArray</code> is a wrapper around many matrix classes in <code>R</code>, including <code>matrix</code>, <code>sparseMatrix</code> and <code>HDF5Array</code>. The last of which allows for out-of-memory computation, which means the computation is done outside of RAM. This will inevitably slow down the computational speed, but the major gain in doing this is that we can perform computations on data much larger than what our RAM can store at once.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/DelayedArray">DelayedArray</a></span><span class="op">)</span></code></pre></div>
<pre><code>## Loading required package: Matrix</code></pre>
<pre><code>## 
## Attaching package: 'Matrix'</code></pre>
<pre><code>## The following object is masked from 'package:S4Vectors':
## 
##     expand</code></pre>
<pre><code>## 
## Attaching package: 'DelayedArray'</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     aperm, apply, rowsum</code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/HDF5Array">HDF5Array</a></span><span class="op">)</span></code></pre></div>
<pre><code>## Loading required package: rhdf5</code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">da_matrix</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedArray-class.html">DelayedArray</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/realize.html">realize</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">logcounts</a></span><span class="op">(</span><span class="va">cellbench</span><span class="op">)</span>, <span class="st">"HDF5Array"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">da_matrix</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "HDF5Matrix"
## attr(,"package")
## [1] "HDF5Array"</code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html">seed</a></span><span class="op">(</span><span class="va">da_matrix</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "HDF5ArraySeed"
## attr(,"package")
## [1] "HDF5Array"</code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">da_output</span> <span class="op">=</span> <span class="fu"><a href="../reference/Cepo.html">Cepo</a></span><span class="op">(</span>exprsMat <span class="op">=</span> <span class="va">da_matrix</span>, cellType <span class="op">=</span> <span class="va">cellbench</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></code></pre></div>
<p>Even though out-of-memory computation is slow, one way that we can speed up the computation is through parallel processing. This requires some configurations of the <code>DelayedArray</code> package via the <code>setAutoBPPARAM</code> function. <code>BiocParallel</code> package uses the <code>MulticoreParam</code> parameter for Linux/Mac and <code>SnowParam</code> for Windows.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel">BiocParallel</a></span><span class="op">)</span>

<span class="va">BPPARAM</span> <span class="op">=</span> <span class="kw">if</span> <span class="op">(</span><span class="va">.Platform</span><span class="op">$</span><span class="va">OS.type</span> <span class="op">==</span> <span class="st">"windows"</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html">SnowParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="op">}</span> <span class="kw">else</span> <span class="op">{</span>
  <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="op">}</span>

<span class="fu">DelayedArray</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/blockApply.html">setAutoBPPARAM</a></span><span class="op">(</span>BPPARAM <span class="op">=</span> <span class="va">BPPARAM</span><span class="op">)</span> <span class="co">## Setting two cores for computation</span>

<span class="va">da_output_parallel</span> <span class="op">=</span> <span class="fu"><a href="../reference/Cepo.html">Cepo</a></span><span class="op">(</span>exprsMat <span class="op">=</span> <span class="va">da_matrix</span>, cellTypes <span class="op">=</span> <span class="va">cellbench</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span>

<span class="fu">DelayedArray</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/blockApply.html">setAutoBPPARAM</a></span><span class="op">(</span>BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html">SerialParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="co">## Revert back to only one core</span></code></pre></div>
</div>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.0.3 (2020-10-10)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Catalina 10.15.7
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] BiocParallel_1.24.1         HDF5Array_1.18.0           
##  [3] rhdf5_2.34.0                DelayedArray_0.16.0        
##  [5] Matrix_1.2-18               UpSetR_1.4.0               
##  [7] SingleCellExperiment_1.12.0 SummarizedExperiment_1.20.0
##  [9] Biobase_2.50.0              GenomicRanges_1.42.0       
## [11] GenomeInfoDb_1.26.2         IRanges_2.24.1             
## [13] S4Vectors_0.28.1            BiocGenerics_0.36.0        
## [15] MatrixGenerics_1.2.0        matrixStats_0.57.0         
## [17] Cepo_0.0.1                  BiocStyle_2.18.1           
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.5                lattice_0.20-41          
##  [3] assertthat_0.2.1          rprojroot_2.0.2          
##  [5] digest_0.6.27             plyr_1.8.6               
##  [7] R6_2.5.0                  evaluate_0.14            
##  [9] ggplot2_3.3.3             pillar_1.4.7             
## [11] sparseMatrixStats_1.2.0   zlibbioc_1.36.0          
## [13] rlang_0.4.10              rmarkdown_2.6            
## [15] pkgdown_1.6.1             labeling_0.4.2           
## [17] splines_4.0.3             textshaping_0.2.1        
## [19] desc_1.2.0                statmod_1.4.35           
## [21] stringr_1.4.0             RCurl_1.98-1.2           
## [23] munsell_0.5.0             compiler_4.0.3           
## [25] xfun_0.20                 pkgconfig_2.0.3          
## [27] systemfonts_0.3.2         htmltools_0.5.0          
## [29] gridExtra_2.3             tibble_3.0.4             
## [31] GenomeInfoDbData_1.2.4    bookdown_0.21            
## [33] crayon_1.3.4              rhdf5filters_1.2.0       
## [35] bitops_1.0-6              grid_4.0.3               
## [37] gtable_0.3.0              lifecycle_0.2.0          
## [39] magrittr_2.0.1            scales_1.1.1             
## [41] stringi_1.5.3             farver_2.0.3             
## [43] reshape2_1.4.4            XVector_0.30.0           
## [45] fs_1.5.0                  limma_3.46.0             
## [47] DelayedMatrixStats_1.12.1 ellipsis_0.3.1           
## [49] ragg_0.4.0                vctrs_0.3.6              
## [51] Rhdf5lib_1.12.0           tools_4.0.3              
## [53] glue_1.4.2                yaml_2.2.1               
## [55] colorspace_2.0-0          BiocManager_1.30.10      
## [57] memoise_1.1.0             knitr_1.30               
## [59] patchwork_1.1.1</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Hani Jieun Kim, Kevin Wang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
